input_output.ipynb

```diff
@@ -0,0 +1,917 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%load_ext tutorial.tests.testsuite"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Input and output in Python"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "---\n",
+    "\n",
+    "## References\n",
+    "- [input and output tutorial](https://docs.python.org/3/tutorial/inputoutput.html) from the official python tutorial.\n",
+    "- [Is explicitly closing files important?](https://stackoverflow.com/questions/7395542/is-explicitly-closing-files-important)\n",
+    "- [context managers](https://docs.python.org/3/library/contextlib.html) from the python standard library\n",
+    "\n",
+    "## Introduction\n",
+    "An important aspect to learn in a new programming language is input/output (or I/O). For our program to be useful, we need to be able to interact with the outside world. This means performing operartions like:\n",
+    "- reading and writing data to a file\n",
+    "- printing text output\n",
+    "- asking for user input\n",
+    "- connecting to databases or other network services\n",
+    "\n",
+    "\n",
+    "The majority of these operations are covered by the python standard library. We are going to see how to use them in this chapter.\n",
+    "\n",
+    "<div class=\"alert alert-block alert-info\">\n",
+    "⚠️ In reference to the chapter on functional programming, it is interesting to note that these functions perform *side-effects*.  Therefore any code containing these operations is no longer *pure* and is not referentially transparent. The same function can return different values for the same argument if called multiple time and the function can have *long-distance* effects. That means that they can modify the state of the program somewhere else leading to unexpected results. \n",
+    "    \n",
+    "Therefore, I always suggest trying to separate input and output from the other computations in your program. For example, if you have a complex calculation requring several user inputs at several stages of the process, consider writing a function that only performs the calculation given all inputs and then require all inputs separately, for example through a single file. This makes your code easier to debug, test and understand.\n",
```

Comment:
Let's use "we suggest", "we propose", etc. This is a collaborative project, after all.

```suggestion
    "Therefore, we suggest separating input and output from the other computations in your program. For example, if you have a complex calculation requiring several user inputs at several stages of the process, consider writing a function that only performs the calculation given all inputs and then requires all inputs separately, for example, through a single file. This makes your code easier to debug, test and understand.\n",
```